### DISCUSSION TITLE
How to contraint the position/orientation of a variable in the optimization?

### DISCUSSION THREAD

**Discussion by Marcel1604** [casadi/casadi#3321] [Q&A] [answered]
Hello together,

I am currently working on an optimization problem with casadi 3.5.5 where I want to constraint the orientation of a variable s to travel above a surface that is represented by a point cloud.

s = opti.variable(N, 6)
p = points of the surface in numpy array format with 6 entries per point.
The first 3 entries represent the position x,y,z, the last three entries the orientation/normal vector with i,j,k. The values of p are prior known in numerical values.

I also introduced a time step dt = 0.1s and a horizon length N = 10. What worked so far is that I implemented in a for-loop (for k in range(N-1):) the kinematic movement of the position. However, what did not work at all is the constraint of the orientation of s, since it should always be orthogonal to the surface and thus be equal to the closest normal vector of the respective point p.

I tried this, to find the closest position of s in the point cloud p and select the corresponding normal vector:
for k in range(N-1):
  differences = p - ca.repmat(s[k+1,:3], p.shape[0], 1)
  squared_norms = differences[:,0]**2 + differences[:,1]**2 + differences[:,2]**2
  squared_norms = squared_norms + 1e-6 # prevent sqrt of 0
  euclidean_norms = ca.sqrt(squared_norms)
  closest_idx = ca.mmin(euclidean_norms) == euclidean_norms
  closest_normal = closest_idx.T@p[3:6]

and this code works, when I check the solving this outputs correct numerical values that make sense.
However, when I now try to add the constraint, that the orientation must always be orthogonal:
  opti.subject_to( s[k+1,3:6] == closest_normal )

or:
  dot_product = ca.dot(s[k+1, 3:6], closest_normal)
  opti.subject_to(dot_product == 1)

I always get the same error:
<class 'RuntimeError'>: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:159:
.../casadi/core/function_internal.cpp:145: Error calling IpoptInterface::init for 'solver':
Error in Function::factory for 'nlp' [MXFunction] at .../casadi/core/function.cpp:1634:
Failed to create nlp_grad:[x, p, lam:f, lam:g]->[f, g, grad:gamma:x, grad:gamma:p] with {"gamma": [f, g]}:
.../casadi/core/factory.hpp:367: Gradient generation failed:
Error in MX::gradient at .../casadi/core/generic_matrix.hpp:1203:
Error in MX::jtimes at .../casadi/core/generic_matrix.hpp:1191:
Error in MX::reverse at .../casadi/core/mx.cpp:1741:
Error in MXFunction::ad_reverse at .../casadi/core/mx_function.cpp:1042:
Error in MX::ad_reverse for node of type N6casadi4MMinE at .../casadi/core/mx.cpp:2044:
.../casadi/core/mx_node.cpp:325: 'ad_reverse' not defined for class N6casadi4MMinE

---
Then I thought about a different approach, that I am not using a time step dt, but let the next poses to be equal to _any_ of the point cloud values p. However, I did not manage to find a formulation that somehow works like this:
s[k+1, :3] == any(p)
Everything I did led to the same error as above..

What is my mistake or why is it not working? I hope you can help me to solve this issue.

Best regards

**Comment by jgillis** [comment, marked as answer]
I believe the missing rule for mmin was added in casadi 3.6, but anyhow you probably het better results with lse (logsumexponential). casadi 3.6 has such operator but you can also implement it yourself

**Reply by Marcel1604** [reply]
okay thank you very much for that advice. 
I already tried to install the latest casadi version via "pip install casadi", but I am restricted to the python v3.8 and so far I could not find a way to run casadi v3.6.3 via python v.3.8. Is there a way to do that?

**Reply by jgillis** [reply]
That is super weird. Can you get some kind of verbose info from pip?

**Comment by Marcel1604** [comment]
thank you very much, it was indeed the casadi version that caused the problem!
